mmaction/datasets/ssn_dataset.py

```diff
@@ -0,0 +1,660 @@
+            proposals = [
+                SSNInstance(
+                    int(x[3]),
+                    int(x[4]),
+                    num_frames,
+                    label=int(x[0]),
+                    best_iou=float(x[1]),
+                    overlap_self=float(x[2])) for x in p[3]
+                if int(x[4]) > int(x[3])
+            ]
+            proposals = list(
+                filter(lambda x: x.start_frame < num_frames, proposals))
+            video_infos.append(
+                dict(
+                    frame_dir=frame_dir,
+                    video_id=p[0],
+                    total_frames=num_frames,
+                    gts=gts,
+                    proposals=proposals))
+        return video_infos
+
+    def evaluate(self, results, metrics, logger):
+        # TODO
+        pass
+
+    def __len__(self):
+        return len(self.video_infos)
+
+    def construct_proposal_pools(self):
+        """Construct positve proposal pool, incomplete proposal pool and
+        background proposal pool of the entire dataset."""
+        for v in self.video_infos:
+            fg = self.get_positives(v['gts'], v['proposals'],
+                                    self.assigner.fg_iou_thr,
+                                    self.sampler.add_gt_as_proposals)
+            self.fg_pool.extend([(v['video_id'], prop) for prop in fg])
+            incomp, bg = self.get_negatives(
+                v['proposals'], self.assigner.incomplete_iou_thr,
+                self.assigner.bg_iou_thr, self.assigner.bg_coverage_thr,
+                self.assigner.incomplete_overlap_thr)
+            self.incomp_pool.extend([(v['video_id'], prop) for prop in incomp])
+            self.bg_pool.extend([v['video_id'], prop] for prop in bg)
+
+    def get_all_gt(self):
+        """Fetch groundtruth instances of the entire dataset."""
+        gt_list = []
+        for video in self.video_infos:
+            vid = video['video_id']
+            gt_list.extend([[
+                vid, gt.label - 1, gt.start_frame / video['total_frames'],
+                gt.end_frame / video['total_frames']
+            ] for gt in video['gts']])
+
+        return gt_list
+
+    def _filter_records(self):
+        """Filter videos with no groundtruth during training."""
+        valid_inds = []
+        for i, x in enumerate(self.video_infos):
+            if len(x['gts']) > 0:
+                valid_inds.append(i)
+
+        return valid_inds
+
+    def get_positives(self, gts, proposals, fg_thresh, with_gt=True):
+        """Get positive/foreground proposals.
+
+        Args:
+            gts (list): List of groundtruth instances.
+            proposals (list): List of proposal instances.
+            fg_thresh (float): Minimum threshold of overlap of
+                positive/foreground proposals and groundtruths.
+            with_gt (bool): Whether include groundtruth instances in
+                positive proposals.
```

Comment:
with_gt (bool): Whether to include groundtruth instances in positive proposals. Default: True.